### PR TITLE
Anchor.FM: Get most recent episode when no GUID is passed.

### LINKED ISSUE #18514: Anchor.fm: Default to latest episode if episode not provided
#### Is your feature request related to a problem? Please describe.

It's possible that users could come from Anchor.fm without an episode specified, particularly from the Dashboard link

image

When that happens, you end up with an empty post, rather than having the podcast episode starter content.

image (1)

#### Describe the solution you'd like

Fallback to the most recent episode in the RSS feed for the episode post content.

### PR DESCRIPTION
**Pull request by pablinos**
Fixes #18514

#### Changes proposed in this Pull Request:
This makes the `anchor_episode` parameter optional. When it's not supplied we'll create the episode post using the most recent episode of the podcast. 

#### Jetpack product discussion

This is part of the Anchor.FM integration project.

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

* Go to `/wp-admin/post-new.php?anchor_podcast=5243218` on your development site, ensuring that you have the beta/experimental features enabled.
* You'll be presented with a blank post.
* With this branch, do the same thing,  and you should see a new post in the editor, populated with the latest episode of the podcast.

#### Proposed changelog entry for your changes:
Not required, as this is a change to the Anchor.FM integration, which is launched.

### CODE REVIEW

File: projects/plugins/jetpack/_inc/lib/class-jetpack-podcast-helper.php [hunk 1]
```diff
@@ -98,10 +98,10 @@ function ( $track ) {
 	/**
 	 * Gets a specific track from the supplied feed URL.
 	 *
-	 * @param string $guid     The GUID of the track.
+	 * @param string $guid     The GUID of the track. Omit or pass a falsey value to get the most recent track.
 	 * @return array|WP_Error  The track object or an error object.
```
Comment (retrofox) [inline projects/plugins/jetpack/_inc/lib/class-jetpack-podcast-helper.php:102]: _**minor**_. Let's make it optional


```suggestion
	 * @param {string=} $guid    The GUID of the track. Omit or pass a falsey value to get the most recent track.
	 * @return {array|WP_Error}  The track object or an error object.
```

**Review by retrofox — COMMENTED** [review]
I've tested and it works as expected.

image

I'd like to get a 👀 from other Explorers folks.
Also, there are some failing tests too.

**Review by retrofox — APPROVED** [review]
LGTM. It works as expected. 🎉

File: projects/plugins/jetpack/_inc/lib/class-jetpack-podcast-helper.php [hunk 2]
```diff
@@ -98,10 +98,10 @@ function ( $track ) {
 	/**
 	 * Gets a specific track from the supplied feed URL.
 	 *
-	 * @param string $guid     The GUID of the track.
-	 * @return array|WP_Error  The track object or an error object.
+	 * @param {string=} $guid    The GUID of the track. Omit or pass a falsey value to get the most recent track.
+	 * @return {array|WP_Error}  The track object or an error object.
 	 */
-	public function get_track_data( $guid ) {
+	public function get_track_data( $guid = null ) {
```
Comment (creativecoder) [inline projects/plugins/jetpack/_inc/lib/class-jetpack-podcast-helper.php:104]: I wouldn't except `get_track_data` to work like this, falling back to the most recent episode--that feels a bit like a "magic" API for the podcast helper that's likely to cause confusion or bugs down the road.

File: projects/plugins/jetpack/extensions/blocks/anchor-fm/anchor-fm.php [hunk 3]
```diff
@@ -105,34 +105,32 @@ function process_anchor_params() {
 		if ( ! \is_wp_error( $rss ) ) {
 			update_post_meta( $post->ID, 'jetpack_anchor_podcast', $podcast_id );
 
-			if ( ! empty( $episode_id ) ) {
```
Comment (creativecoder) [inline projects/plugins/jetpack/extensions/blocks/anchor-fm/anchor-fm.php:108]: Why not add logic outside this conditional for when there isn't an episode id that get's the most recent episode using the helper?

```php
if ( empty( $episode_id ) ) {
    ...
```

**Review by creativecoder — COMMENTED** [review]
I'm not a fan of shimming in a default case to `get_track_data`. I think it would be better to handle this imperatively, by getting the player data and checking for the most recent track when there's no episode id.

If nothing else, I think it's clearer what's going on when you skim the code.

File: projects/plugins/jetpack/_inc/lib/class-jetpack-podcast-helper.php (hunk 2 above)
Comment (pablinos) [reply projects/plugins/jetpack/_inc/lib/class-jetpack-podcast-helper.php:104]: Yeah fair enough, I probably was being a bit lazy! I'll rework it.

File: projects/plugins/jetpack/_inc/lib/class-jetpack-podcast-helper.php [hunk 4]
```diff
@@ -98,8 +98,8 @@ function ( $track ) {
 	/**
 	 * Gets a specific track from the supplied feed URL.
 	 *
-	 * @param string $guid     The GUID of the track.
-	 * @return array|WP_Error  The track object or an error object.
+	 * @param string $guid    The GUID of the track.
+	 * @return {array|WP_Error}  The track object or an error object.
```
Comment (creativecoder) [inline projects/plugins/jetpack/_inc/lib/class-jetpack-podcast-helper.php:102]: I might just revert the changes from this file to keep the changeset cleaner (non blocking).

File: projects/plugins/jetpack/extensions/blocks/anchor-fm/anchor-fm.php [hunk 5]
```diff
@@ -105,10 +105,18 @@ function process_anchor_params() {
...
+				if ( $latest_episode ) {
+					$episode_id = $latest_episode->get_id();
+				}
+			}
+
 			if ( ! empty( $episode_id ) ) {
 				$track = $podcast_helper->get_track_data( $episode_id );
 				if ( ! \is_wp_error( $track ) ) {
-					update_post_meta( $post->ID, 'jetpack_anchor_episode', $episode_id );
+					update_post_meta( $post->ID, 'jetpack_anchor_episode', $track['guid'] );
```
Comment (creativecoder) [inline projects/plugins/jetpack/extensions/blocks/anchor-fm/anchor-fm.php:119]: Any specific reason you swapped `$track['guid']` to `$episode_id`?

(Intuitively, it makes sense... store the thing you just used to look up an episode for when you need to do the same lookup in the future. I'm just curious if you found this made a difference in some way.)

**Review by creativecoder — APPROVED** [review]
✅ Looks good, works as expected when passing only `anchor_episode` and not an episode id.

File: projects/plugins/jetpack/extensions/blocks/anchor-fm/anchor-fm.php (hunk 5 above)
Comment (pablinos) [reply projects/plugins/jetpack/extensions/blocks/anchor-fm/anchor-fm.php:119]: No, it came from when I was using a default parameter with `get_track_data`, but as you say, it struck me as more intuitive.

**Review by creativecoder — APPROVED** [review]
Re-review ✅

**Review by retrofox — COMMENTED** [review]
LGTM

**Review by retrofox — APPROVED** [review]
LGTM

**Comment by pablinos** [comment]
r219999-wpcom

**Pull request merged**